dom/augdom/src/testing.rs

```diff
@@ -0,0 +1,255 @@
+        by_placeholder_text PlaceholderText
+    }
+
+    strat_method! {
+        /// Find by aria `role`.
+        ///
+        /// The default choice for interactive elements like buttons.
+        by_role             Role
+    }
+
+    strat_method! {
+        /// Find by element's normalized text content.
+        by_text             Text
+    }
+
+    strat_method! {
+        /// Find by form element's current `value`.
+        by_display_value    DisplayValue
+    }
+
+    strat_method! {
+        /// Find by `img`'s `alt` attribute.
+        by_alt_text         AltText
+    }
+
+    strat_method! {
+        /// Find by `title` attribute's or svg `title` tag's normalized text content.
+        by_title            Title
+    }
+
+    strat_method! {
+        /// Find by `data-testid` attribute. Not visible to humans, only
+        /// use as a last resort.
+        by_test_id          TestId
+    }
+}
+
+/// The final description of a subtree query. The methods on this struct
+/// execute the underlying search and return the results in various forms.
+pub struct Found<'find, 'pat, 'node, N> {
+    strat: Strategy,
+    pattern: &'pat str,
+    finder: &'find Finder<'node, N>,
+}
+
+impl<'find, 'pat, 'node, N> Found<'find, 'pat, 'node, N>
+where
+    N: Dom,
+{
+    /// Wrap the query in a [`MutationObserver`] with async methods that resolve
+    /// once the wrapped query could succeed or a large timeout has expired.
+    ///
+    /// [`MutationObserver`]: https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver
+    pub fn until(&self) -> Until<'_, 'find, 'pat, 'node, N> {
+        Until::new(self)
+    }
+
+    /// Execute the query and return the only matching node in the queried
+    /// subtree.
+    ///
+    /// # Panics
+    ///
+    /// If more than one matching node is found.
+    pub fn one(&self) -> Option<N> {
+        let mut matches = self.many().into_iter();
+        let first = matches.next();
+        assert!(matches.next().is_none(), "`one()` returned more than one matching node");
+        first
+    }
+
+    /// Execute the query and return a `Vec` of matching nodes in the queried
+    /// subtree.
+    pub fn many(&self) -> Vec<N> {
+        // first accumulate the subtree
+        let mut candidates = Vec::new();
+        collect_children_dfs_preorder(self.finder.target, &mut candidates);
+
+        // then keep only those which match
+        candidates.retain(|n| self.matches(n));
+
+        candidates
+    }
+
+    fn normalize(s: impl AsRef<str>) -> String {
+        s.as_ref().split_whitespace().collect::<Vec<_>>().join(" ")
+    }
+
+    fn matches(&self, node: &N) -> bool {
+        use Strategy::*;
+        match self.strat {
+            Text => Some(node.get_inner_text()),
+            LabelText => node
+                .get_attribute("id")
+                .map(|id| {
+                    let selector = format!("label[for={}]", id);
```

Comment:
Does this actually find labels in cases when they're "next to" the intended element?